File: back/hooks/posttooluse/sensors.go
// Package main is the AIDOS PostToolUse sensors hook — the feedback wall and the
// twin of the wall (CLAUDE.md §6, KRD §74: "PostToolUse = les sensors"). After
// each agent diff below the waterline, it resolves the CHANGED CODE set, runs the
// computational sensor suite (gofmt, vet, lint, archtest, affected), and emits a
// verdict: on any failing check `block` with an actionable BlockReason (code
// SENSOR_FAILED); on all-green `allow`. Every run is recorded append-only in
// runtime.sensor_runs.
//
// DETERMINISM-FIRST (CLAUDE.md §6/§8): the verdict is a pure function of the event
// and the code it points at — Aggregate has no clock, no rng, no I/O, so the same
// results always yield the same decision (the rapid property is its reproducibility
// mirror). The hook INVOKES the frozen tools (gofmt, go vet, go test) as
// subprocesses; it never reimplements them, and it never delegates the verdict to
// an LLM (a sensor judged by an agent would be a determinism gap).
//
// The sensors check CONFORMANCE to the kernel; they do not define it. The hook
// writes nothing above the waterline — sensor_runs is a runtime audit log below the
// line (ADR 0014).
package main

// Verdict is the hook's decision. There are exactly two — there is no third
// verdict (the rapid invariant pins this; the DB CHECK constraint enforces it).
type Verdict string

const (
	// VerdictAllow lets the diff proceed (all sensors green).
	VerdictAllow Verdict = "allow"
	// VerdictBlock refuses the diff (at least one sensor failed). on_fail: block.
	VerdictBlock Verdict = "block"
)

// BlockCode is the stable, machine-readable code of a refusal.
type BlockCode string

// CodeSensorFailed is the one code this hook emits: at least one computational
// sensor failed on the changed code.
const CodeSensorFailed BlockCode = "SENSOR_FAILED"

// BlockReason is the actionable refusal shape shared across AIDOS block sites
// (CLAUDE.md §2: code, severity, explanation, how_to_fix[]). It mirrors the S04
// wall BlockReason JSON contract; each hook declares the shape locally (the
// existing two-hook precedent — pretooluse + ci-ratchet). A block without a
// BlockReason becomes a prison; the refusal always names the door.
type BlockReason struct {
	Code        BlockCode `json:"code"`
	Severity    string    `json:"severity"`
	Explanation string    `json:"explanation"`
	HowToFix    []string  `json:"how_to_fix"`
}

// CheckResult is one computational sensor's outcome on the changed code.
//   - Pass:    the sensor is green. An errored result MUST have Pass=false.
//   - Errored: the sensor could not produce a verdict (crashed / missing). This is
//     a FAILURE made explicit (KRD §82 .passthrough() anti-pattern), never ignored.
//   - Output:  the captured tool output (for the BlockReason / audit log).
//   - DurationMS: wall-clock duration, for the /sensors panel.
type CheckResult struct {
	Name       string `json:"name"`
	Pass       bool   `json:"pass"`
	Errored    bool   `json:"errored,omitempty"`
	Output     string `json:"output,omitempty"`
	DurationMS int64  `json:"duration_ms"`
}

// Decision is the aggregator's output: a verdict, the failing check names (exactly
// the failing subset, sorted, no drop), and — on block — the BlockReason.
type Decision struct {
	Verdict     Verdict      `json:"verdict"`
	Failing     []string     `json:"failing,omitempty"`
	BlockReason *BlockReason `json:"block_reason,omitempty"`
}

// Aggregate is the deep, well-named heart of on_fail:block: block IFF at least one
// check failed, allow otherwise. Pure and total. A result is "failing" iff !Pass
// (an errored result has Pass=false, so it counts as failing — never silently
// dropped). The failing[] list is exactly the failing subset, in input order.
func Aggregate(results []CheckResult) Decision {
	var failing []string
	for _, r := range results {
		if !r.Pass {
			failing = append(failing, r.Name)
		}
	}
	if len(failing) == 0 {
		return Decision{Verdict: VerdictAllow}
	}
	return Decision{
		Verdict:     VerdictBlock,
		Failing:     failing,
		BlockReason: sensorFailedBlockReason(failing, results),
	}
}

// sensorFailedBlockReason builds the canonical actionable refusal naming the
// failing sensor(s) and how to fix them.
func sensorFailedBlockReason(failing []string, results []CheckResult) *BlockReason {
	first := failing[0]
	explanation := "Sensor en échec sur le code changé : « " + first + " »"
	if len(failing) > 1 {
		explanation += " (et " + itoa(len(failing)-1) + " autre(s))"
	}
	explanation += ". L'agent ne s'auto-certifie que sur le computationnel ; le diff est bloqué tant qu'un capteur est rouge (on_fail: block, KRD §74)."
	// Surface the first failing sensor's output, when present, to make the refusal
	// concrete (the audit log carries the full per-check output).
	if out := outputFor(first, results); out != "" {
		explanation += " Sortie : " + truncate(out, 400)
	}
	return &BlockReason{
		Code:        CodeSensorFailed,
		Severity:    "error",
		Explanation: explanation,
		HowToFix: []string{
			"Reproduisez localement le capteur fautif sur le paquet changé (gofmt -l / go vet / go test).",
			"Inspectez le BlockReason avec `aidos explain SENSOR_FAILED` et la route Workbench /sensors.",
			"Réparez le diff (red→green) avant le prochain tool-call ; le mur de feedback tient tant que le capteur reste rouge.",
		},
	}
}

func outputFor(name string, results []CheckResult) string {
	for _, r := range results {
		if r.Name == name {
			return r.Output
		}
	}
	return ""
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "…"
}

// itoa avoids pulling strconv into the hot type file for a single small int.
func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

File: front/web/lib/sensors.ts
/**
 * The sensors — the Workbench /sensors projection source (AIDOS step S07).
 *
 * THE FEEDBACK WALL (CLAUDE.md §6, KRD §74: "PostToolUse = les sensors"): after
 * each agent diff below the waterline the PostToolUse Go hook
 * (back/hooks/posttooluse) resolves the CHANGED CODE set, runs the computational
 * sensor suite — gofmt, vet, lint, archtest, affected — and emits a verdict:
 * `block` on any failing check (BlockReason, code SENSOR_FAILED), `allow` on all
 * green. Every run is recorded append-only in runtime.sensor_runs.
 *
 * This module holds only the DECLARED, static description of that suite — the
 * canonical check list (field-for-field with adapters.go's CanonicalChecks), a
 * representative latest run, and a representative SENSOR_FAILED block event — so
 * the /sensors panel shows exactly what the Go hook runs and the refusal it emits.
 * One source, no drift.
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): a static, declared registry (no clock, no
 * rng, no I/O). It mirrors the Go suite's order and codes; the reproducibility
 * mirror lib/sensors.test.ts pins the check set, the canonical code, and the
 * pure verdict (aggregate ⇒ block iff any check failed).
 *
 * READ-ONLY: /sensors is a visualization; it projects runtime.sensor_runs, it does
 * not re-encode it. It writes no truth and exposes no capability — the sensors are
 * enforced by the harness-invoked PostToolUse hook, never from a screen.
 * ui-completeness is vacuously satisfied: there is no headless capability hidden
 * here, there is none (the rerun-on-demand MCP is OQ-S07-1, not built yet).
 */

/** The one block code the PostToolUse sensors emit (matches sensors.go). */
export const SENSOR_FAILED = "SENSOR_FAILED" as const;

/** The two verdicts — there is no third (matches sensors.go / the DB CHECK). */
export type Verdict = "allow" | "block";

/**
 * One computational sensor in the suite. `slot` is the KRD §74 role it fills
 * (typecheck/format → lint → archtest → affected); `tool` names the frozen Go
 * tool the adapter invokes (it never reimplements it).
 */
export interface Sensor {
	/** The check name, as recorded in sensor_runs (matches CanonicalChecks). */
	name: string;
	/** The KRD §74 role this check fills. */
	slot: string;
	/** The frozen Go tool the adapter invokes. */
	tool: string;
	/** One-line role in AIDOS terms (FR-first per ADR 0011). */
	role: string;
}

/**
 * The computational sensor suite, in KRD §74 order (typecheck/format → lint →
 * archtest → affected). Mirrors adapters.go `CanonicalChecks`
 * = ["gofmt","vet","lint","archtest","affected"] field-for-field.
 */
export const SENSOR_SUITE: readonly Sensor[] = [
	{
		name: "gofmt",
		slot: "typecheck / format",
		tool: "gofmt -l",
		role: "le code changé est-il au format canonique ? Une liste non vide (gofmt -l) est un échec.",
	},
	{
		name: "vet",
		slot: "typecheck",
		tool: "go vet",
		role: "vet statique sur les paquets affectés — les pièges que le compilateur tolère mais que vet refuse.",
	},
	{
		name: "lint",
		slot: "lint",
		tool: "go build",
		role: "porte Go-strict : compile les paquets affectés (go-arch-lint/depguard pour un lint riche = OQ-S07-2).",
	},
	{
		name: "archtest",
		slot: "archtest",
		tool: "boundary check (in-process)",
		role: "une projection sous la ligne ne doit jamais importer l'arbre de vérité back/kernel/** (ADR 0002/0014).",
	},
	{
		name: "affected",
		slot: "run-mirrors --affected",
		tool: "go test",
		role: "les tests / miroirs affectés (go test sur les paquets changés) — la preuve exécutable reste verte.",
	},
] as const;

/** One sensor's outcome in a recorded run (projects a sensor_check_results row). */
export interface CheckResult {
	name: string;
	pass: boolean;
	/** An errored check (crashed / missing) is a FAILURE made explicit (KRD §82). */
	errored?: boolean;
	durationMs: number;
}

/** The shared actionable-refusal shape (matches sensors.go BlockReason). */
export interface BlockReason {
	code: typeof SENSOR_FAILED;
	severity: "error";
	explanation: string;
	howToFix: readonly string[];
}

/** A recorded sensor run (projects a runtime.sensor_runs row + its children). */
export interface SensorRun {
	runId: string;
	/** The changed-code set the run scoped to (the affected packages / files). */
	target: string;
	verdict: Verdict;
	results: readonly CheckResult[];
	/** Present iff verdict === "block". */
	blockReason?: BlockReason;
}

/**
 * Aggregate — the pure heart of on_fail:block, mirroring sensors.go `Aggregate`:
 * block IFF at least one check failed, allow otherwise. A check is failing iff
 * `!pass` (an errored check has pass=false, so it counts — never silently
 * dropped). The reproducibility mirror (lib/sensors.test.ts) pins it against the
 * Go invariant: same results → same verdict, exactly the failing subset.
 */
export function aggregate(results: readonly CheckResult[]): {
	verdict: Verdict;
	failing: string[];
} {
	const failing = results.filter((r) => !r.pass).map((r) => r.name);
	return { verdict: failing.length === 0 ? "allow" : "block", failing };
}

/**
 * A representative LATEST run — all sensors green, verdict allow. Shaped like a
 * real sensor_runs projection (the panel reads the table; this is the seeded
 * sample so /sensors renders before live wiring lands, OQ-S07-3).
 */
export const SAMPLE_LATEST_RUN: SensorRun = {
	runId: "run-0007-clean",
	target: "back/hooks/posttooluse (paquet changé)",
	verdict: "allow",
	results: [
		{ name: "gofmt", pass: true, durationMs: 12 },
		{ name: "vet", pass: true, durationMs: 188 },
		{ name: "lint", pass: true, durationMs: 240 },
		{ name: "archtest", pass: true, durationMs: 3 },
		{ name: "affected", pass: true, durationMs: 1024 },
	],
} as const;

/**
 * A representative SENSOR_FAILED block event — the BlockReason the PostToolUse
 * hook returns when a sensor goes red on the changed code. Verbatim-shaped from
 * sensorFailedBlockReason in sensors.go, so the panel shows a true refusal, not an
 * invented one. The how_to_fix always names how to make the diff green again.
 */
export const SAMPLE_BLOCK_EVENT: SensorRun = {
	runId: "run-0006-affected-red",
	target: "back/hooks/posttooluse (paquet changé)",
	verdict: "block",
	results: [
		{ name: "gofmt", pass: true, durationMs: 11 },
		{ name: "vet", pass: true, durationMs: 176 },
		{ name: "lint", pass: true, durationMs: 233 },
		{ name: "archtest", pass: true, durationMs: 3 },
		{ name: "affected", pass: false, durationMs: 842 },
	],
	blockReason: {
		code: SENSOR_FAILED,
		severity: "error",
		explanation:
			"Sensor en échec sur le code changé : « affected ». L'agent ne s'auto-certifie que sur le computationnel ; le diff est bloqué tant qu'un capteur est rouge (on_fail: block, KRD §74). Sortie : go test failed: --- FAIL: TestAggregate_BlocksOnAnyFailure (0.00s).",
		howToFix: [
			"Reproduisez localement le capteur fautif sur le paquet changé (gofmt -l / go vet / go test).",
			"Inspectez le BlockReason avec `aidos explain SENSOR_FAILED` et la route Workbench /sensors.",
			"Réparez le diff (red→green) avant le prochain tool-call ; le mur de feedback tient tant que le capteur reste rouge.",
		],
	},
} as const;

/** Fresh copy of the sensor suite, in canonical order. */
export function sensorSuite(): Sensor[] {
	return SENSOR_SUITE.map((s) => ({ ...s }));
}
